mypy/subtypes.py

```diff
@@ -143,10 +143,15 @@ def visit_instance(self, left: Instance) -> bool:
             rname = right.type.fullname()
             if not left.type.has_base(rname) and rname != 'builtins.object':
                 return False
-
             # Map left type to corresponding right instances.
             t = map_instance_to_supertype(left, right.type)
-
+            # Check that one of the types does not have type args or
+            # the number of type args is the same and
+            # each left type arg is acceptable in place of the matching right one
+            if not t.args or not right.args:
+                return True
+            if len(t.args) != len(right.args):
+                return False
```

Comment:
At this point (after mapping instance to supertype) they should always have same numbers of arguments, because ``t`` and ``right`` are now ``Instance``s of the same ``TypeInfo``. You could put ``assert  len(t.args) == len(right.args) == len(right.type.defn.type_vars)`` and see what happens. If the assert is triggered, then it means there is a bug somewhere else.